1) Requirements Setup 

In [109]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import yfinance as yf

import statsmodels.api as sm
from statsmodels.stats.diagnostic import acorr_ljungbox
from scipy.stats import binomtest as binom_test

from datetime import datetime, timedelta
from pathlib import Path
from scipy.stats import t
import os
from scipy import stats
from statsmodels.tsa.stattools import adfuller



2) Getting the relevant data from the Dataset 

In [110]:
# getting the data from ELia's dataset 
# paths 
Root_dir = os.getcwd()
Data_dir = os.path.join(Root_dir, "data_extraction", "raw_df")

# load the CSV + computing the log returns
def load_with_log_returns(csv_filename,lag = 1):
	"""
	Load a CSV from the Data_dir and compute log returns and also the Log Prices  on the 'Close' column.
	This function is robust: it will handle CSVs that have either 'Date' or 'Datetime' (or both).
	It raises a clear error if neither date column or the 'Close' column is present.
	"""
	path = os.path.join(Data_dir, csv_filename)
	# Read first without forcing parse_dates so we can detect which date columns exist
	df = pd.read_csv(path)
	# Parse available date columns robustly
	for col in ['Date', 'Datetime']:
		if col in df.columns:
			df[col] = pd.to_datetime(df[col], errors='coerce')
	# Prefer 'Date' as the index, otherwise use 'Datetime' if present
	if 'Date' in df.columns and not df['Date'].isna().all():
		df.set_index('Date', inplace=True)
	elif 'Datetime' in df.columns and not df['Datetime'].isna().all():
		df.set_index('Datetime', inplace=True)
	else:
		raise ValueError(f"CSV {path} must contain a 'Date' or 'Datetime' column. Found: {list(df.columns)}")
	# Ensure 'Close' exists
	if 'Close' not in df.columns:
		raise ValueError(f"CSV {path} must contain a 'Close' column. Found: {list(df.columns)}")
	# Compute log returns safely (first entry will be NaN)
	df['Log_Returns'] = np.log(df['Close'] / df['Close'].shift(lag))
	
	df['Log_Prices'] = np.log(df['Close'])
	return df





nasdaq_daily_df = load_with_log_returns('nasdaq_daily_df.csv')
nasdaq_weekly_df = load_with_log_returns('nasdaq_weekly_df.csv')
nasdaq_monthly_df = load_with_log_returns('nasdaq_monthly_df.csv')
nasdaq_hourly_df = load_with_log_returns('nasdaq_hourly_df.csv')

# The DataFrame uses 'Date' (or 'Datetime') as the index, so select the 'Log_Returns' column directly.
nasdaq_daily_log_returns = nasdaq_daily_df['Log_Returns'].dropna()
nasdaq_weekly_log_returns = nasdaq_weekly_df['Log_Returns'].dropna()
nasdaq_monthly_log_returns = nasdaq_monthly_df['Log_Returns'].dropna()
nasdaq_hourly_df_log_returns = nasdaq_hourly_df['Log_Returns'].dropna()

nasdaq_hourly_df_log_prices = nasdaq_hourly_df['Log_Prices'].dropna()
nasdaq_daily_df_log_prices = nasdaq_daily_df['Log_Prices'].dropna()
nasdaq_weekly_df_log_prices = nasdaq_weekly_df['Log_Prices'].dropna()
nasdaq_monthly_df_log_prices = nasdaq_monthly_df['Log_Prices'].dropna()


3) Testing the Martingale Properties of the Log Return 

In [111]:

def test_mean_returns(returns):
    """
    Basically a variant of the T-test where we test if the mean return is equal to zero for each respective frequency.

    Returns:
        (mean_return, std_return, se_return, t_stat, p_value_super, p_value_sub, p_value_two)

    p_value_super: one-sided p-value for H1: mean < 0 (supermartingale)
    p_value_sub:   one-sided p-value for H1: mean > 0 (submartingale)
    p_value_two:   two-sided p-value for H1: mean != 0
    """
    # converts it all to a big ol numpy and drops the NaNs
    r = returns.values if hasattr(returns, "values") else np.array(returns)
    r = np.asarray(r)
    r = r[~np.isnan(r)]

    n = len(r)
    if n < 2:
        raise ValueError("Not enough data points to perform the test.")

    mean_return = np.mean(r)
    std_return = np.std(r, ddof=1)
    se_return = std_return / np.sqrt(n)

    # handle zero-variance case --> it just a likkle safety net 
    if se_return == 0:
        t_stat = 0.0
    else:
        t_stat = mean_return / se_return

    # one-sided p-values 
    p_value_super = t.cdf(t_stat, df=n-1)   
    p_value_sub = 1 - p_value_super        
    p_value_two = 2 * min(p_value_super, p_value_sub)  

    return mean_return, std_return, se_return, t_stat, p_value_super, p_value_sub, p_value_two

def print_mean_return_results(returns, label="Mean Return"):
    (mean, std, se, t_stat,
     p_super, p_sub, p_two) = test_mean_returns(returns)

    n = len(returns)

    print("\n==============================")
    print(f"TEST – {label}")
    print("==============================")
    print(f"Number of obs:                    {n}")
    print(f"Mean return:                      {mean:.6e}")
    print(f"Std dev:                          {std:.6e}")
    print(f"Std error of mean:                {se:.6e}")
    print(f"t-statistic:                      {t_stat:.3f}")
    print(f"\nTwo-sided p-value (H1: mean ≠ 0): {p_two:.4g}")
    print(f"One-sided p-value (H1: mean < 0 → supermartingale): {p_super:.4g}")
    print(f"One-sided p-value (H1: mean > 0 → submartingale):   {p_sub:.4g}")

    # Interpretation
    print("\nInterpretation:")
    if p_two < 0.05:
        if mean > 0:
            print("→ Significant *positive* drift: SUBMARTINGALE behaviour.")
        else:
            print("→ Significant *negative* drift: SUPERMARTINGALE behaviour.")
    else:
        print("→ Mean return not significantly different from zero: MARTINGALE-compatible.")

print_mean_return_results(nasdaq_daily_log_returns,   "NASDAQ – DAILY log returns")
print_mean_return_results(nasdaq_weekly_log_returns,  "NASDAQ – WEEKLY log returns")
print_mean_return_results(nasdaq_monthly_log_returns, "NASDAQ – MONTHLY log returns")
print_mean_return_results(nasdaq_hourly_df_log_returns, "NASDAQ – HOURLY log returns")




TEST – NASDAQ – DAILY log returns
Number of obs:                    3992
Mean return:                      5.747994e-04
Std dev:                          1.297311e-02
Std error of mean:                2.053283e-04
t-statistic:                      2.799

Two-sided p-value (H1: mean ≠ 0): 0.005144
One-sided p-value (H1: mean < 0 → supermartingale): 0.9974
One-sided p-value (H1: mean > 0 → submartingale):   0.002572

Interpretation:
→ Significant *positive* drift: SUBMARTINGALE behaviour.

TEST – NASDAQ – WEEKLY log returns
Number of obs:                    828
Mean return:                      2.775642e-03
Std dev:                          2.598318e-02
Std error of mean:                9.029778e-04
t-statistic:                      3.074

Two-sided p-value (H1: mean ≠ 0): 0.002182
One-sided p-value (H1: mean < 0 → supermartingale): 0.9989
One-sided p-value (H1: mean > 0 → submartingale):   0.001091

Interpretation:
→ Significant *positive* drift: SUBMARTINGALE behaviour.

TEST – NASDAQ

4.1 CW ratio for the log returns 

In [ ]:
# Strong random Walk Hypothesis (RW1)
# Apply the Cowles and Jones (1937) test to check for independence of returns.
# The test compares the frequency of sequences (two successive returns with same sign)
# to the frequency of reversals (two successive returns with opposite signs).

def cowles_jones_test(returns, name=""):
    """
    Apply the Cowles and Jones (1937) test for independence of returns.

    Parameters:
        returns : array-like or pandas Series of returns
        name    : optional label for printing

    Returns:
        dict with keys: NS, NR, CJ, z_stat, p_value
    """
    # prepare data
    r = returns.dropna().values if hasattr(returns, "dropna") else np.asarray(returns)
    non_zero_mask = r != 0
    r = r[non_zero_mask]

    if len(r) < 2:
        raise ValueError("Not enough observations (after removing zeros/NaNs) to perform the Cowles and Jones test.")

    signs = np.sign(r)
    prod = signs[:-1] * signs[1:]

    NS = np.sum(prod > 0)
    NR = np.sum(prod < 0)
    N = NS + NR

    if NR == 0 or N == 0:
        raise ValueError("Not enough variability in returns to perform the Cowles and Jones test.")

    CJ_hat = NS / NR

    # Under RW1 with mu = 0: P(sequence) = P(reversal) = 0.5
    p_hat = NS / N
    p0 = 0.5
    se = np.sqrt(p0 * (1 - p0) / N)
    z_stat = (p_hat - p0) / se

    # use normal distribution for z-test
    from scipy.stats import norm
    p_value = 2 * (1 - norm.cdf(abs(z_stat)))

    print(f"\n===== Cowles & Jones Test ({name}) =====")
    print(f"N_S (sequences, same sign) : {NS}")
    print(f"N_R (reversals, opp. sign) : {NR}")
    print(f"Total pairs N              : {N}")
    print(f"CĴ = N_S / N_R            : {CJ_hat:.4f}")
    print(f"p̂ = N_S / (N_S+N_R)       : {p_hat:.4f}")
    print(f"z-stat (H0: p = 0.5)       : {z_stat:.4f}")
    print(f"p-value (two-sided)        : {p_value:.6f}")
    if p_value < 0.05:
        print("→ Reject RW1 (μ=0) at 5% level.")
    else:
        print("→ Cannot reject RW1 (μ=0) at 5% level.")

    return {
        "NS": int(NS),
        "NR": int(NR),
        "CJ": float(CJ_hat),
        "z_stat": float(z_stat),
        "p_value": float(p_value)
        }

cj_daily   = cowles_jones_test(nasdaq_daily_log_returns,        "Daily log returns")
cj_hourly  = cowles_jones_test(nasdaq_hourly_df_log_returns,    "Hourly log returns")
cj_weekly  = cowles_jones_test(nasdaq_weekly_log_returns,       "Weekly log returns")
cj_monthly = cowles_jones_test(nasdaq_monthly_log_returns,      "Monthly log returns")



===== Cowles & Jones Test (Daily log returns) =====
N_S (sequences, same sign) : 1994
N_R (reversals, opp. sign) : 1996
Total pairs N              : 3990
CĴ = N_S / N_R            : 0.9990
p̂ = N_S / (N_S+N_R)       : 0.4997
z-stat (H0: p = 0.5)       : -0.0317
p-value (two-sided)        : 0.974741
→ Cannot reject RW1 (μ=0) at 5% level.

===== Cowles & Jones Test (Hourly log returns) =====
N_S (sequences, same sign) : 763
N_R (reversals, opp. sign) : 764
Total pairs N              : 1527
CĴ = N_S / N_R            : 0.9987
p̂ = N_S / (N_S+N_R)       : 0.4997
z-stat (H0: p = 0.5)       : -0.0256
p-value (two-sided)        : 0.979584
→ Cannot reject RW1 (μ=0) at 5% level.

===== Cowles & Jones Test (Weekly log returns) =====
N_S (sequences, same sign) : 424
N_R (reversals, opp. sign) : 402
Total pairs N              : 826
CĴ = N_S / N_R            : 1.0547
p̂ = N_S / (N_S+N_R)       : 0.5133
z-stat (H0: p = 0.5)       : 0.7655
p-value (two-sided)        : 0.443987
→ Cannot reject RW1 

4.2 CW ratio for the log prices

In [113]:
# Run Cowles & Jones on price or log-price series (fixed implementation)
def cowles_jones_test_prices(series, name="", is_log=True):
    """
    Cowles & Jones test applied to a series of prices or log-prices.
    
    Parameters:
        series : pandas Series or array-like (price levels or log-prices)
        name   : optional label for printing
        is_log : if True, treat `series` as log-prices and compute returns as differences;
                 if False, treat `series` as price levels and compute percentage returns.
    
    Returns:
        dict with keys: NS, NR, CJ, z_stat, p_value
    """
    # ensure pandas is available in cell scope
    s = series.dropna() if hasattr(series, "dropna") else pd.Series(series).dropna()
    p = s.values
    # remove exact zeros (relevant for level prices)
    non_zero_mask = p != 0
    p = p[non_zero_mask]
    if len(p) < 2:
        raise ValueError("Not enough observations (after removing zeros/NaNs) to perform the Cowles and Jones test.")
    if is_log:
        # log-returns are differences of log-prices
        returns = np.diff(p)
    else:
        # percentage returns for price levels
        returns = np.diff(p) / p[:-1]
    signs = np.sign(returns)
    prod = signs[:-1] * signs[1:]
    NS = int(np.sum(prod > 0))
    NR = int(np.sum(prod < 0))
    N = int(NS + NR)
    if NR == 0 or N == 0:
        raise ValueError("Not enough variability in returns to perform the Cowles and Jones test.")
    CJ_hat = NS / NR
    p_hat = NS / N
    p0 = 0.5
    se = np.sqrt(p0 * (1 - p0) / N)
    z_stat = (p_hat - p0) / se
    from scipy.stats import norm
    p_value = 2 * (1 - norm.cdf(abs(z_stat)))
    print(f"\n===== Cowles & Jones Test on Series ({name}) =====")
    print(f"N_S (sequences, same sign) : {NS}")
    print(f"N_R (reversals, opp. sign) : {NR}")
    print(f"Total pairs N              : {N}")
    print(f"CĴ = N_S / N_R            : {CJ_hat:.4f}")
    print(f"p̂ = N_S / (N_S+N_R)       : {p_hat:.4f}")
    print(f"z-stat (H0: p = 0.5)       : {z_stat:.4f}")
    print(f"p-value (two-sided)        : {p_value:.6f}")
    if p_value < 0.05:
        print("→ Reject RW1 (μ=0) at 5% level.")
    else:
        print("→ Cannot reject RW1 (μ=0) at 5% level.")
    return {
        "NS": int(NS),
        "NR": int(NR),
        "CJ": float(CJ_hat),
        "z_stat": float(z_stat),
        "p_value": float(p_value)
    }

# The precomputed variables `nasdaq_*_df_log_prices` are Series of log-prices (they were created above),
# so pass them directly and set is_log=True to compute differences
cj_daily_log_prices   = cowles_jones_test_prices(nasdaq_daily_df_log_prices,   "Daily log prices",   is_log=True)
cj_hourly_log_prices  = cowles_jones_test_prices(nasdaq_hourly_df_log_prices,  "Hourly log prices",  is_log=True)
cj_weekly_log_prices  = cowles_jones_test_prices(nasdaq_weekly_df_log_prices,  "Weekly log prices",  is_log=True)
cj_monthly_log_prices = cowles_jones_test_prices(nasdaq_monthly_df_log_prices, "Monthly log prices", is_log=True)


===== Cowles & Jones Test on Series (Daily log prices) =====
N_S (sequences, same sign) : 1993
N_R (reversals, opp. sign) : 1996
Total pairs N              : 3989
CĴ = N_S / N_R            : 0.9985
p̂ = N_S / (N_S+N_R)       : 0.4996
z-stat (H0: p = 0.5)       : -0.0475
p-value (two-sided)        : 0.962115
→ Cannot reject RW1 (μ=0) at 5% level.

===== Cowles & Jones Test on Series (Hourly log prices) =====
N_S (sequences, same sign) : 763
N_R (reversals, opp. sign) : 764
Total pairs N              : 1527
CĴ = N_S / N_R            : 0.9987
p̂ = N_S / (N_S+N_R)       : 0.4997
z-stat (H0: p = 0.5)       : -0.0256
p-value (two-sided)        : 0.979584
→ Cannot reject RW1 (μ=0) at 5% level.

===== Cowles & Jones Test on Series (Weekly log prices) =====
N_S (sequences, same sign) : 424
N_R (reversals, opp. sign) : 401
Total pairs N              : 825
CĴ = N_S / N_R            : 1.0574
p̂ = N_S / (N_S+N_R)       : 0.5139
z-stat (H0: p = 0.5)       : 0.8008
p-value (two-sided)        : 0.

In [114]:
# Testing Drift in 1 period returns (each frequency)

def drift_ttest(returns, name = "", alternative = "two-sided"):

    r = returns.dropna().values if hasattr(returns, "dropna") else np.asarray(returns)
    r = np.asarray(r)
    r = r[~np.isnan(r)]

    n = len(r)
    if n < 2:
        raise ValueError("Not enough observations to perform t-test (need at least 2).")

    mean_r = r.mean()
    std_r = r.std(ddof=1)
    se_r = std_r / np.sqrt(n)

    # coz we ave tha some dif 0 n tha
    if se_r == 0:
        t_stat = 0.0
    else:
        t_stat = mean_r / se_r
    df = n - 1

    if alternative == "greater":
        # H1: mean > 0
        p_value = 1 - stats.t.cdf(t_stat, df=df)
    elif alternative == "less":
        # H1: mean < 0
        p_value = stats.t.cdf(t_stat, df=df)
    else:
        # two-sided
        # use survival function for numerical stability
        p_value = 2 * min(stats.t.cdf(t_stat, df=df), 1 - stats.t.cdf(t_stat, df=df))

    print(f"\n===== Drift t-test (1-period, {name}) =====")
    print(f"n          : {n}")
    print(f"mean(r)    : {mean_r:.6e}")
    print(f"std(r)     : {std_r:.6e}")
    print(f"t-stat     : {t_stat:.4f}")
    print(f"p-value    : {p_value:.6f} ({alternative})")
    if p_value < 0.05:
        print("→ Reject H0: evidence of drift.")
    else:
        print("→ Cannot reject H0: mean not significantly different from 0.")
    return {"mean": mean_r, "std": std_r, "t_stat": t_stat, "p_value": p_value}

# 1-period drift tests
drift_ttest(nasdaq_hourly_df['Log_Returns'],  "Hourly log returns",  alternative="greater")
drift_ttest(nasdaq_daily_df['Log_Returns'],   "Daily log returns",   alternative="greater")
drift_ttest(nasdaq_weekly_df['Log_Returns'],  "Weekly log returns",  alternative="greater")
drift_ttest(nasdaq_monthly_df['Log_Returns'], "Monthly log returns", alternative="greater")


===== Drift t-test (1-period, Hourly log returns) =====
n          : 1528
mean(r)    : 1.090797e-04
std(r)     : 5.680545e-03
t-stat     : 0.7506
p-value    : 0.226501 (greater)
→ Cannot reject H0: mean not significantly different from 0.

===== Drift t-test (1-period, Daily log returns) =====
n          : 3992
mean(r)    : 5.747994e-04
std(r)     : 1.297311e-02
t-stat     : 2.7994
p-value    : 0.002572 (greater)
→ Reject H0: evidence of drift.

===== Drift t-test (1-period, Weekly log returns) =====
n          : 828
mean(r)    : 2.775642e-03
std(r)     : 2.598318e-02
t-stat     : 3.0739
p-value    : 0.001091 (greater)
→ Reject H0: evidence of drift.

===== Drift t-test (1-period, Monthly log returns) =====
n          : 190
mean(r)    : 1.241308e-02
std(r)     : 4.943746e-02
t-stat     : 3.4610
p-value    : 0.000333 (greater)
→ Reject H0: evidence of drift.


{'mean': np.float64(0.012413083223296532),
 'std': np.float64(0.04943745639891111),
 't_stat': np.float64(3.4609900423081084),
 'p_value': np.float64(0.000332670743255159)}

(4.3) Campbell, Lo MacKinlay (1997) 

In [115]:

def long_horizon_returns(log_price_series, h):
    lp = np.asarray(log_price_series.dropna())

    n_blocks = len(lp) // h
    if n_blocks <= 1:
        raise ValueError(f"Not enough data for horizon h={h}.")

    lp_trunc = lp[:n_blocks * h]
    lp_start = lp_trunc[0::h]
    lp_end   = lp_trunc[h-1::h]
    r_h = lp_end - lp_start
    return r_h


# 2. Campbell long-horizon drift t-test
def long_horizon_drift_ttest(log_price_series, name="", horizons=(1,5,20,60)):
    print(f"\n===== Campbell Long-Horizon Drift Test ({name}) =====")
    for h in horizons:
        try:
            r_h = long_horizon_returns(log_price_series, h)
        except ValueError as e:
            print(f"h={h}: {e}")
            continue
        
        n = len(r_h)
        mean_r = r_h.mean()
        std_r = r_h.std(ddof=1)
        se_r = std_r / np.sqrt(n)
        t_stat = mean_r / se_r if se_r != 0 else 0.0
        df = n - 1
        p_value = 1 - stats.t.cdf(t_stat, df=df)  # one-sided (greater)

        print(f"\nh = {h} steps")
        print(f"  n_blocks : {n}")
        print(f"  mean R^{h} : {mean_r:.6e}")
        print(f"  std(R^{h}) : {std_r:.6e}")
        print(f"  t-stat : {t_stat:.4f}")
        print(f"  p-value (H0: mean=0, greater) : {p_value:.6f}")
        if p_value < 0.05:
            print("  → Reject H0: drift visible at this horizon.")
        else:
            print("  → Cannot reject H0 at this horizon.")


# 3. Make sure your data has a Log_Price column
nasdaq_hourly_df["Log_Price"]  = np.log(nasdaq_hourly_df["Close"])
nasdaq_daily_df["Log_Price"]   = np.log(nasdaq_daily_df["Close"])
nasdaq_weekly_df["Log_Price"]  = np.log(nasdaq_weekly_df["Close"])
nasdaq_monthly_df["Log_Price"] = np.log(nasdaq_monthly_df["Close"])


# 4. Run the Campbell-style long-horizon drift tests
long_horizon_drift_ttest(nasdaq_hourly_df["Log_Price"],  "Hourly log prices",  horizons=(1,5,20,60))
long_horizon_drift_ttest(nasdaq_daily_df["Log_Price"],   "Daily log prices",   horizons=(1,5,20,60))
long_horizon_drift_ttest(nasdaq_weekly_df["Log_Price"],  "Weekly log prices",  horizons=(1,5,20,60))
long_horizon_drift_ttest(nasdaq_monthly_df["Log_Price"], "Monthly log prices", horizons=(1,5,20,60))



===== Campbell Long-Horizon Drift Test (Hourly log prices) =====

h = 1 steps
  n_blocks : 1529
  mean R^1 : 0.000000e+00
  std(R^1) : 0.000000e+00
  t-stat : 0.0000
  p-value (H0: mean=0, greater) : 0.500000
  → Cannot reject H0 at this horizon.

h = 5 steps
  n_blocks : 305
  mean R^5 : 5.949643e-04
  std(R^5) : 1.095416e-02
  t-stat : 0.9486
  p-value (H0: mean=0, greater) : 0.171801
  → Cannot reject H0 at this horizon.

h = 20 steps
  n_blocks : 76
  mean R^20 : 2.691898e-03
  std(R^20) : 2.232551e-02
  t-stat : 1.0511
  p-value (H0: mean=0, greater) : 0.148283
  → Cannot reject H0 at this horizon.

h = 60 steps
  n_blocks : 25
  mean R^60 : 7.353976e-03
  std(R^60) : 3.185076e-02
  t-stat : 1.1544
  p-value (H0: mean=0, greater) : 0.129842
  → Cannot reject H0 at this horizon.

===== Campbell Long-Horizon Drift Test (Daily log prices) =====

h = 1 steps
  n_blocks : 3993
  mean R^1 : 0.000000e+00
  std(R^1) : 0.000000e+00
  t-stat : 0.0000
  p-value (H0: mean=0, greater) : 0.500


# **RW1 Testing Results – Interpretation and Link to the Martingale Property**

Under **RW1**, the log-price process satisfies:

[
X_t = X_{t-1} + \varepsilon_t, \qquad \mathbb{E}[\varepsilon_t] = 0,
]

which implies:

[
\mathbb{E}[X_t \mid \mathcal{F}*{t-1}] = X*{t-1}
\quad \Longleftrightarrow \quad
\mathbb{E}[r_t] = 0.
]

Therefore, **RW1 is equivalent to log-returns being a martingale difference sequence (MDS)**.
Testing RW1 means empirically checking:

1. **Is the mean return zero?**
2. **Are signs of returns random (no persistence)?**
3. **Does drift appear only at long horizons (Campbell decomposition)?**

Each test captures a different implication of the martingale property.

---

# **1. Drift t-Test on Log Returns**

This tests:

[
H_0: \mathbb{E}[r_t] = 0 \quad \text{(martingale / RW1)}
]

A significantly positive mean return implies:

* **Submartingale**: (\mathbb{E}[r_t] > 0)
* **Supermartingale**: (\mathbb{E}[r_t] < 0)

### **Results by Frequency**

| Frequency   | Mean Return     | t-stat | p-value | Interpretation                       |
| ----------- | --------------- | ------ | ------- | ------------------------------------ |
| **Hourly**  | Not significant | 0.75   | 0.2265  | **Martingale-compatible** (no drift) |
| **Daily**   | Significant > 0 | 2.80   | 0.0026  | **Submartingale** (positive drift)   |
| **Weekly**  | Significant > 0 | 3.07   | 0.0011  | **Submartingale**                    |
| **Monthly** | Significant > 0 | 3.46   | 0.0003  | **Submartingale**                    |

**Summary:**

* At **hourly** frequency, the data behaves like a **martingale**.
* At **daily → monthly** frequencies, significant **positive drift** appears.
* This is consistent with empirical finance: short horizons ≈ martingale, longer horizons ≈ positive risk premium.

---

# **2. Cowles & Jones Directional Test (Sign Test)**

This evaluates whether return signs follow a fair Bernoulli(0.5) process, testing:

[
H_0: P(\text{up}) = P(\text{down}) = 1/2.
]

It checks for **predictability in direction of returns** (violation of martingale).

### **Results for Log Returns**

| Frequency   | p-value | Interpretation    |
| ----------- | ------- | ----------------- |
| **Hourly**  | 0.9796  | Cannot reject RW1 |
| **Daily**   | 0.9747  | Cannot reject RW1 |
| **Weekly**  | 0.4440  | Cannot reject RW1 |
| **Monthly** | 0.1670  | Cannot reject RW1 |

### **Results for Log Prices**

Same conclusion as for returns: **all p-values > 0.16**, so **no evidence against RW1**.

**Summary:**
Direction of returns is **indistinguishable from random** across all frequencies → **no sign predictability**, consistent with a martingale.

---

# **3. Campbell Long-Horizon Drift Tests**

Campbell–Lo–MacKinlay show that:

* For **RW1**, drift should appear **only at long horizons** as returns compound.
* For **short horizons**, drift is extremely small and often statistically indistinguishable from zero.

This test checks:

[
R_t^{(h)} = X_{t+h} - X_t \quad \text{(non-overlapping (h)-period log returns)}
]

and performs:

[
H_0: \mathbb{E}[R_t^{(h)}] = 0.
]

### **Results by Frequency**

#### **Hourly Log Prices**

* Drift **never significant**, even at 60-hour horizon.
* Consistent with a **martingale**.

#### **Daily Log Prices**

* Drift becomes significant from **h = 5, 20, 60**.
* Long-horizon drift emerges, consistent with **long-run compounding**.

#### **Weekly Log Prices**

* Significant drift at **h = 5, 20, 60**.

#### **Monthly Log Prices**

* Very strong drift from **h = 5, 20, 60**.

**Summary:**

* **Short horizon (1-period)** → log-prices behave like a **martingale**.
* **Long horizon** → drift becomes visible (positive risk premium).
* This matches Campbell’s theory that drift is only reliably detectable at large (h).

---

# **Overall RW1 Interpretation**

### **Hourly Frequency**

* No significant drift
* No directional predictability
* No long-horizon cumulative drift
* **Fully consistent with a martingale (RW1).**

### **Daily Frequency**

* Significant positive drift
* But **no sign predictability**
* Long-horizon drift emerges
* **Not RW1, but consistent with a submartingale with small drift.**

### **Weekly & Monthly Frequency**

* Increasingly strong positive drift
* No directional predictability
* Strong long-horizon drift
* **Clear submartingale behavior**, but still no forecasting power in signs.

---

# **Link to the Martingale Property and Predictability**

A process is a **martingale in returns** if:

[
\mathbb{E}[r_{t+1} \mid \mathcal{F}_t] = 0.
]

From the RW1 results:

### ✔ **Hourly data → Martingale-compatible**

* No drift
* No directional predictability
* No long-horizon drift

Thus **no arbitrage prediction is possible**, and pricing via martingale methods is fully justified.

### ✔ **Daily to Monthly data → Submartingale, not pure martingale**

* Drift (> 0) → expected return positive
* But signs remain unpredictable
* This indicates a **risk premium**, not forecastability
* Still consistent with **efficient markets** (no predictable short-term excess returns).

---

# **Final Takeaway**

* **RW1 holds only at very short horizons** (hourly).
* As the horizon increases, **positive drift becomes statistically detectable**, but
  **return signs remain unpredictable**, so the market is still **efficient**.
* This pattern is exactly what asset-pricing theory predicts:

  * small horizons → martingale-like
  * long horizons → drift (risk premium)
  * no evidence of systematic predictability at any frequency.




(5) RW2 - Second Hypothesis - "Random Walk with Drift & Uncorrelated Innovations" --> Trivially Strong Random Walk

Under RW2, the log-price process is assumed to follow an --> Xt = μ + Xt+1 + εt  where the innovation is centered 

∴ this HYP assumes --> log prices contain a unit root and --> increments of ΔXt are stationary with mean μ (do proof in document)
--> we can say that the prices follow a random walk with drift.

Run a little DF test --> Test Unit root --> H0 = Unit Root (random walk) and H1 = Stationary 

In [120]:
import numpy as np
import pandas as pd
import statsmodels.api as sm
from statsmodels.tsa.stattools import adfuller
from scipy import stats

def adf_test_log_prices(log_price_series, name=""):
    # Convert to pandas Series
    x = pd.Series(log_price_series.dropna())
    
    print(f"\n\n===== RW2 TEST for {name} =====")
    print(f"Observations: {len(x)}")
    
    # ADF Test with the constant ----> gonna test if its a RW with drift
    print("\n-- ADF Test with Constant (Random Walk with Drift) --")
    adf_c = adfuller(x, regression='c', autolag='AIC')
    stat_c, p_c, usedlag_c, nobs_c, crit_c, _ = adf_c
    
    print(f"ADF Statistic : {stat_c:.4f}")
    print(f"p-value       : {p_c:.6f}")
    print(f"Lags used     : {usedlag_c}")
    print("Critical values:")
    for k, v in crit_c.items():
        print(f"  {k}% : {v:.4f}")
        
    if p_c < 0.05:
        print("→ Reject H0: log-price is stationary (NO unit root).")
    else:
        print("→ Cannot reject H0: log-price has a unit root (RW2 compatible).")
    
    
    # ADF test with constant + Trend ----> gonna test if its a RW with drift and trend
    print("\n-- ADF Test with Constant + Trend (RW with Drift and Trend) --")
    adf_ct = adfuller(x, regression='ct', autolag='AIC')
    stat_ct, p_ct, usedlag_ct, nobs_ct, crit_ct, _ = adf_ct
    
    print(f"ADF Statistic : {stat_ct:.4f}")
    print(f"p-value       : {p_ct:.6f}")
    print(f"Lags used     : {usedlag_ct}")
    print("Critical values:")
    for k, v in crit_ct.items():
        print(f"  {k}% : {v:.4f}")
        
    if p_ct < 0.05:
        print("→ Reject H0: trend-stationary (NO unit root).")
    else:
        print("→ Cannot reject H0: stochastic trend (RW2/RW3 compatible).")
    
    
    
    dx = x.diff().dropna()
    x_lag = x.shift(1).dropna()
    y = dx.loc[x_lag.index]
    
    X_no_trend = sm.add_constant(x_lag)
    df_no_trend = sm.OLS(y, X_no_trend).fit()
    
    print("\n-- DF Regression Without Trend: ΔX_t = α + φ X_{t-1} + ε_t --")
    print(df_no_trend.summary())
    
    phi = df_no_trend.params[1]
    phi_se = df_no_trend.bse[1]
    phi_t = df_no_trend.tvalues[1]
    phi_p = df_no_trend.pvalues[1]
    
    print("\nKey DF Stats (no trend):")
    print(f"φ estimate       : {phi:.6f}")
    print(f"Std. error       : {phi_se:.6f}")
    print(f"t-statistic      : {phi_t:.4f}")
    print(f"p-value          : {phi_p:.6f}")
    
    if phi_p < 0.05 and phi < 0:
        print("→ Reject unit root: NOT a random walk.")
    else:
        print("→ Cannot reject unit root (RW2-compatible).")
    
    

    t_index = np.arange(len(x_lag))
    trend = pd.Series(t_index, index=x_lag.index)
    
    X_trend = np.column_stack((np.ones(len(x_lag)), trend.values, x_lag.values))
    df_trend = sm.OLS(y.values, X_trend).fit()
    
    alpha_ct, beta_ct, phi_ct = df_trend.params
    phi_se_ct = df_trend.bse[2]
    phi_t_ct = df_trend.tvalues[2]
    phi_p_ct = df_trend.pvalues[2]
    
    print("\n-- DF Regression With Trend: ΔX_t = α + β t + φ X_{t-1} + ε_t --")
    print(df_trend.summary())
    
    print("\nKey DF Stats (with trend):")
    print(f"Trend coefficient β : {beta_ct:.6f}")
    print(f"φ coefficient        : {phi_ct:.6f}")
    print(f"Std. error φ         : {phi_se_ct:.6f}")
    print(f"t-statistic φ        : {phi_t_ct:.4f}")
    print(f"p-value φ            : {phi_p_ct:.6f}")
    
    if phi_p_ct < 0.05 and phi_ct < 0:
        print("→ Reject unit root (trend-stationary).")
    else:
        print("→ Cannot reject unit root (stochastic trend).")
    
    
    return {
        "ADF_const_p": p_c,
        "ADF_trend_p": p_ct,
        "phi_no_trend": phi,
        "phi_trend": phi_ct,
        "phi_no_trend_p": phi_p,
        "phi_trend_p": phi_p_ct
    }


# Run RW2 tests for all frequencies
rw2_hourly  = adf_test_log_prices(nasdaq_hourly_df["Log_Price"],  "Hourly log prices")
rw2_daily   = adf_test_log_prices(nasdaq_daily_df["Log_Price"],   "Daily log prices")
rw2_weekly  = adf_test_log_prices(nasdaq_weekly_df["Log_Price"],  "Weekly log prices")
rw2_monthly = adf_test_log_prices(nasdaq_monthly_df["Log_Price"], "Monthly log prices")




===== RW2 TEST for Hourly log prices =====
Observations: 1529

-- ADF Test with Constant (Random Walk with Drift) --
ADF Statistic : -0.4969
p-value       : 0.892629
Lags used     : 24
Critical values:
  1%% : -3.4347
  5%% : -2.8635
  10%% : -2.5678
→ Cannot reject H0: log-price has a unit root (RW2 compatible).

-- ADF Test with Constant + Trend (RW with Drift and Trend) --
ADF Statistic : -2.0692
p-value       : 0.563314
Lags used     : 24
Critical values:
  1%% : -3.9648
  5%% : -3.4134
  10%% : -3.1288
→ Cannot reject H0: stochastic trend (RW2/RW3 compatible).

-- DF Regression Without Trend: ΔX_t = α + φ X_{t-1} + ε_t --
                            OLS Regression Results                            
Dep. Variable:              Log_Price   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.000
Method:                 Least Squares   F-statistic:                    0.5001
Date:                Wed, 19 Nov 2025   Prob (F

In [119]:
from statsmodels.tsa.stattools import adfuller

def print_adf_lags(series, name):
    x = series.dropna()
    result = adfuller(x, autolag='AIC')
    used_lags = result[2]
    aic = result[5]
    print(f"{name}: ADF AIC-selected lags = {used_lags}, AIC = {aic:.4f}")

print_adf_lags(nasdaq_hourly_df["Log_Price"],  "Hourly")
print_adf_lags(nasdaq_daily_df["Log_Price"],   "Daily")
print_adf_lags(nasdaq_weekly_df["Log_Price"],  "Weekly")
print_adf_lags(nasdaq_monthly_df["Log_Price"], "Monthly")

Hourly: ADF AIC-selected lags = 24, AIC = -11288.3205
Daily: ADF AIC-selected lags = 26, AIC = -23263.2110
Weekly: ADF AIC-selected lags = 4, AIC = -3614.6548
Monthly: ADF AIC-selected lags = 0, AIC = -556.3303


Testing RW3 --> Weak random Walk HYP --> under this hypothesis we assume 

Xt​=μ+Xt−1​+εt​ where the shocks have a mean of 0, constant variance and are uncorrelated across time. --> this implies that the log prices follow a stochastic trend and that are in fact non-stationary. --> we have to look at the returns which are simply the first difference of logs and hence are assumed t be stationary and hence i turn exhnbit zero autocorrealtion 

This is the justification of testing RW3 on log returns and not prices. 


What can be found:

Essentially if we are looking to see if returns are stationary 



In [ ]:
import numpy as np
import pandas as pd
import statsmodels.api as sm
from statsmodels.stats.diagnostic import acorr_ljungbox
from statsmodels.tsa.stattools import acf
from scipy import stats


# RW3 TESTS: Autocorrelation, Ljung-Box Q, Variance Ratio


def variance_ratio_test(r, q=2):
    """
    Computes the Lo–MacKinlay Variance Ratio VR(q).
    Under RW3, VR(q) = 1.
    """
    r = r.dropna().values
    T = len(r)
    mu = np.mean(r)

    # Variance of 1-step returns
    var_1 = np.sum((r - mu)**2) / (T - 1)

    # Variance of q-step aggregated returns
    r_q = pd.Series(r).rolling(q).sum().dropna().values
    var_q = np.sum((r_q - np.mean(r_q))**2) / (len(r_q) - 1)

    VR = var_q / (q * var_1)
    
    # Asymptotic test statistic (homoscedastic case)
    z_stat = (VR - 1) / np.sqrt(2 * (2*q - 1) / (3*q*T))
    p_value = 2 * (1 - stats.norm.cdf(abs(z_stat)))

    return VR, z_stat, p_value


def rw3_tests(return_series, name="", lags_LB=10, VR_list=[2,5,10]):
    """
    Runs all RW3 tests:
    1) Autocorrelation checks
    2) Ljung–Box Q statistic
    3) Variance ratio tests 
    """
    r = return_series.dropna()

    print("\n================================================")
    print(f" RW3 TESTS – {name}")
    print("================================================")
    print(f"Observations: {len(r)}")

    
    # 1. Autocorrelation Function ACF
    
    print("\n--- Autocorrelations (first 10 lags) ---")
    acf_vals = acf(r, nlags=10, fft=False)
    for lag in range(1, 11):
        print(f"ACF lag {lag:2d}: {acf_vals[lag]:.4f}")

    
    # 2. Ljung–Box Q-stat for joint autocor
  
    print("\n--- Ljung–Box Q Test (joint autocorrelation) ---")
    LB = acorr_ljungbox(r, lags=[lags_LB], return_df=True)
    Q = LB['lb_stat'].values[0]
    pQ = LB['lb_pvalue'].values[0]
    print(f"Ljung–Box Q({lags_LB}) = {Q:.4f}, p-value = {pQ:.6f}")

    if pQ < 0.05:
        print("→ Reject RW3: returns show autocorrelation.")
    else:
        print("→ Cannot reject RW3: returns look like white noise.")

   
    #3 Variance-Ratio Tests VR(q)
  
    print("\n--- Variance Ratio Tests (Lo–MacKinlay) ---")
    for q in VR_list:
        VR, zstat, pvr = variance_ratio_test(r, q=q)
        print(f"VR({q}) = {VR:.4f} | z = {zstat:.4f} | p = {pvr:.6f}")
        if pvr < 0.05:
            print("   → Reject RW3 at this horizon.")
        else:
            print("   → Cannot reject RW3.")

    print("\n================================================\n")
    
rw3_tests(nasdaq_hourly_df["Log_Returns"],  "Hourly returns")
rw3_tests(nasdaq_daily_df["Log_Returns"],   "Daily returns")
rw3_tests(nasdaq_weekly_df["Log_Returns"],  "Weekly returns")
rw3_tests(nasdaq_monthly_df["Log_Returns"], "Monthly returns")



 RW3 TESTS – Hourly returns
Observations: 1528

--- Autocorrelations (first 10 lags) ---
ACF lag  1: 0.0345
ACF lag  2: -0.0030
ACF lag  3: -0.0079
ACF lag  4: -0.0436
ACF lag  5: -0.0389
ACF lag  6: -0.0576
ACF lag  7: 0.0314
ACF lag  8: -0.0466
ACF lag  9: -0.0331
ACF lag 10: 0.0733

--- Ljung–Box Q Test (joint autocorrelation) ---
Ljung–Box Q(10) = 27.0792, p-value = 0.002530
→ Reject RW3: returns show autocorrelation.

--- Variance Ratio Tests (Lo–MacKinlay) ---
VR(2) = 1.0352 | z = 1.3761 | p = 0.168796
   → Cannot reject RW3.
VR(5) = 1.0293 | z = 1.0469 | p = 0.295147
   → Cannot reject RW3.
VR(10) = 0.9069 | z = -3.2338 | p = 0.001222
   → Reject RW3 at this horizon.



 RW3 TESTS – Daily returns
Observations: 3992

--- Autocorrelations (first 10 lags) ---
ACF lag  1: -0.0954
ACF lag  2: 0.0544
ACF lag  3: -0.0470
ACF lag  4: -0.0256
ACF lag  5: 0.0012
ACF lag  6: -0.0500
ACF lag  7: 0.0657
ACF lag  8: -0.0732
ACF lag  9: 0.0757
ACF lag 10: -0.0199

--- Ljung–Box Q Test (joint 